flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/PromptHandler.java

```diff
@@ -0,0 +1,188 @@
+ * and the end (right prompt) of each line.
+ */
+public class PromptHandler {
+    private static final char ESCAPE_BACKSLASH = '\\';
+    private static final Map<Character, String> DATE_TIME_FORMATS;
+    private static final ThreadLocalCache<String, SimpleDateFormat> FORMATTER_CACHE =
+            new ThreadLocalCache<String, SimpleDateFormat>() {
+                @Override
+                public SimpleDateFormat getNewInstance(String key) {
+                    return new SimpleDateFormat(key, Locale.ROOT);
+                }
+            };
+
+    static {
+        Map<Character, String> map = new HashMap<>();
+        map.put('D', "yyyy-MM-dd HH:mm:ss.SSS");
+        map.put('m', "mm");
+        map.put('o', "MM");
+        map.put('O', "MMM");
+        map.put('P', "aa");
+        map.put('r', "hh:mm");
+        map.put('R', "HH:mm");
+        map.put('s', "ss");
+        map.put('w', "d");
+        map.put('W', "E");
+        map.put('y', "YY");
+        map.put('Y', "YYYY");
+        DATE_TIME_FORMATS = Collections.unmodifiableMap(map);
+    }
+
+    private static final StyleResolver STYLE_RESOLVER = new StyleResolver(s -> "");
+
+    private final String sessionId;
+    private final Executor executor;
+    private final Supplier<Terminal> terminalSupplier;
+
+    public PromptHandler(String sessionId, Executor executor, Supplier<Terminal> terminalSupplier) {
+        this.sessionId = sessionId;
+        this.executor = executor;
+        this.terminalSupplier = terminalSupplier;
+    }
+
+    public String getPrompt() {
+        return buildPrompt(
+                executor.getSessionConfig(sessionId).get(SqlClientOptions.PROMPT),
+                SqlClientOptions.PROMPT.defaultValue());
+    }
+
+    public String getRightPrompt() {
+        return buildPrompt(
+                executor.getSessionConfig(sessionId).get(SqlClientOptions.RIGHT_PROMPT),
+                SqlClientOptions.RIGHT_PROMPT.defaultValue());
+    }
+
+    private String buildPrompt(String pattern, String defaultValue) {
+        AttributedStringBuilder promptStringBuilder = new AttributedStringBuilder();
+        try {
+            for (int i = 0; i < pattern.length(); i++) {
+                final char c = pattern.charAt(i);
+                if (c == ESCAPE_BACKSLASH) {
+                    if (i == pattern.length() - 1) {
+                        continue;
+                    }
+                    final char nextChar = pattern.charAt(i + 1);
+                    switch (nextChar) {
+                        case 'D':
+                        case 'm':
+                        case 'o':
+                        case 'O':
+                        case 'P':
+                        case 'r':
+                        case 'R':
+                        case 's':
+                        case 'w':
+                        case 'W':
+                        case 'y':
+                        case 'Y':
+                            promptStringBuilder.append(
```

Comment:
this will be applied for letters with leading `\` i.e. to have catalog it should be `\c`.
To have `default_catalogdefault_database` it should be `\c\d`.
However it is also possible to put something between e.g.`\c:\d` or `\c_\d` and etc.